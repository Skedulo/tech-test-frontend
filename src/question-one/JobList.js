import React from 'react';

const Job = ({name, start, end, contact}) => <div className="job">
    <div>{name}</div>
    <div>{start} - {end}</div>
    <div>{contact.name}</div>
</div>

const JobList = ({ jobs }) => <ul>
    {jobs.map(job => <li key={job.id}><Job {...job} /></li>)}
</ul>

export default JobList;