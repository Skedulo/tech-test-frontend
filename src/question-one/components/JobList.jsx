import React from 'react'
import PropTypes from 'prop-types'

import useLoadJobs from '../hooks/useLoadJobs'

import JobItem from './JobItem'

const JobList = ({ searchString$, service }) => {
  const { jobs, isLoading, isInitial } = useLoadJobs(searchString$, service.getJobsWithSearchTerm)

  if (isInitial) {
    return null
  }

  if (isLoading) {
    return 'loading'
  }

  if (!jobs.length) {
    return 'Not Found'
  }

  return (
    <React.Fragment>
      {jobs.map((item) => <JobItem key={item.id} item={item} />)}
    </React.Fragment>
  )
}

JobList.propTypes = {
  searchString$: PropTypes.object.isRequired,
  service: PropTypes.shape({
    getJobsWithSearchTerm: PropTypes.func.isRequired
  }).isRequired
}

export default JobList