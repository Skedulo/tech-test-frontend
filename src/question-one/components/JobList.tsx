import React from 'react'

import useLoadJobs from '../hooks/useLoadJobs'

import JobItem from './JobItem'
import { Subscribable, ObservableInput } from 'rxjs'
import {Job} from '../../types/Job'

type JobListProps =  {
  searchString$: Subscribable<string>,
  service: {
    getJobsWithSearchTerm: () => ObservableInput<Job[]>
  }
}

const JobList: React.FC<JobListProps> = ({ searchString$, service }) => {
  const { jobs, isLoading, isInitial } = useLoadJobs(searchString$, service.getJobsWithSearchTerm)

  if (isInitial) {
    return null
  }

  if (isLoading) {
    return <>loading</>
  }

  if (!jobs.length) {
    return <>Not Found</>
}

  return (
    <React.Fragment>
      {jobs.map((item) => <JobItem key={item.id} item={item} />)}
    </React.Fragment>
  )
}

export default JobList
