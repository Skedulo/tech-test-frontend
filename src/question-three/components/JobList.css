.jobList {
  display: grid;
  grid-auto-flow: row;
  grid-template-columns: 1fr;
  align-content: start;
  grid-gap: var(--spacing);
  padding: var(--spacing);
  overflow: auto;
  background: var(--lighterGray);
}
