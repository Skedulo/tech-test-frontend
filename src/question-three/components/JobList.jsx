import React, {useState, useEffect} from 'react'
import JobCard from './JobCard';
import {from} from 'rxjs';

import './JobList.css';

const INITIAL_STATE = {
    jobs: [],
    isLoading: true
}

const useLoadJobs =  (searchFn) => {
    let [state, updateState] = useState(INITIAL_STATE);

    useEffect(() => {
        let subscription = from(searchFn())
            .subscribe((jobs) => {
                updateState({
                    jobs,
                    isLoading: false
                })
            })

        return () => {
            subscription.unsubscribe()
        }
    }, [searchFn])

    return state;
}

export default ({service, className}) => {
    
  const {jobs, isLoading} = useLoadJobs(service.getJobsWithSearchTerm);

    return (
        <div className={`jobList ${className}`}>
            {jobs.map(job => (
                <JobCard job={job} key={job.id} />
            ))}
        </div>
    )
}