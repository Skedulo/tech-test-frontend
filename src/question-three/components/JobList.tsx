import React, { useState, useEffect } from 'react'

import JobCard from './JobCard'
import { from } from 'rxjs'

import './JobList.css'
import { IJobService } from '../../service/DataService'
import { Job } from '../../types/Job'

interface State {
  jobs : Job[],
  isLoading: Boolean
}

const INITIAL_STATE = {
  jobs: [],
  isLoading: true
}

const useLoadJobs = (searchFn: IJobService["getJobsWithSearchTerm"]) => {
  const [state, updateState] = useState<State>(INITIAL_STATE)

  useEffect(() => {
    const subscription = from(searchFn())
      .subscribe((jobs) => {
        updateState({
          jobs,
          isLoading: false
      })
      })

    return () => {
      subscription.unsubscribe()
    }
  }, [searchFn])

  return state
}

const JobList : React.FC<{service: IJobService, className: string}> = ({ service, className }) => {
  const { jobs, isLoading } = useLoadJobs(service.getJobsWithSearchTerm)

  return (
    <div className={`jobList ${className}`}>
      {isLoading ? 'Loading' : jobs.map(job => (
        <JobCard job={job} key={job.id} />
      ))}
    </div>
  )
}


export default JobList
