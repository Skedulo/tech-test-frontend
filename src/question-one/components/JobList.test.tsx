import React from 'react'
import { render, act } from '@testing-library/react'

import { BehaviorSubject } from 'rxjs'

import JobList from './JobList'

const renderJobList = (initialSearchString = '') => {
  const mockSearchFn = jest.fn()
  const mockSerivce = {
    getJobsWithSearchTerm: mockSearchFn
  }

  const input$ = new BehaviorSubject(initialSearchString)

  let jobList = render(<JobList service={mockSerivce} searchString$={input$}/>)

  return {
    jobList,
    input$,
    mockSearchFn
  }
}

describe('JobList', () => {
  it('should render nothing when useLoadJobs return inital', () => {
    const { jobList } = renderJobList()
    expect(jobList.baseElement.textContent).toBe('')

    jobList.unmount()
  })

  it('should render loading state when useLoadJobs return isLoading', async () => {
    const { jobList } = renderJobList('123')

    expect(jobList.baseElement.textContent).toBe('loading')

    jobList.unmount()
  })

  it('should render \'Not Found\' when useLoadJobs return empty jobs', async () => {
    const { jobList, mockSearchFn } = renderJobList('123')
    mockSearchFn.mockImplementationOnce(() => [[]])

    expect(jobList.baseElement.textContent).toBe('loading')

    await act(async () => {
      await new Promise(resolve => setTimeout(resolve, 600))
    })
    expect(mockSearchFn).toBeCalledWith('123')

    expect(jobList.baseElement.textContent).toBe('Not Found')

    jobList.unmount()
  })

  it('should render matchSnapshot when useLoadJobs return array jobs', async () => {
    const { jobList, mockSearchFn } = renderJobList('123')
    mockSearchFn.mockImplementationOnce(() => [[{ id: '1' }, { id: '2' }]])
    
    expect(jobList.baseElement.textContent).toBe('loading')

    await act(async () => {
      await new Promise(resolve => setTimeout(resolve, 600))
    })

    expect(mockSearchFn).toBeCalledWith('123')
    
    expect(jobList.getAllByTitle(/^Job Item:/)).toHaveLength(2)

    jobList.unmount()
  })
})
