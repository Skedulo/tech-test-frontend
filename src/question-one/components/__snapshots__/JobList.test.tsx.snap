// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobList should render matchSnapshot when useLoadJobs return array jobs 1`] = `
Array [
  "<JobItem/>",
  "<JobItem/>",
]
`;
